File: examples/ffi/swift/Rings.swift
import Foundation

#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

public let ringsSignatureLength = 65

public enum RingsLogLevel: Int32 {
    case debug = 0
    case info = 1
    case warn = 2
    case error = 3
    case trace = 4
}

public enum RingsFfiError: Error, CustomStringConvertible {
    case loadLibrary(String)
    case loadSymbol(String)
    case signerCapacity
    case signerLength(Int)
    case providerCreation
    case request(String)
    case invalidUtf8
    case invalidResponse(String)
    case timeout(String)

    public var description: String {
        switch self {
        case .loadLibrary(let message): return "failed to load Rings library: \(message)"
        case .loadSymbol(let symbol): return "Rings library does not export \(symbol)"
        case .signerCapacity: return "all Swift FFI signer callback slots are in use"
        case .signerLength(let length): return "signer returned \(length) bytes; expected 65"
        case .providerCreation: return "rings_node_new_provider_with_callback returned NULL"
        case .request(let method): return "Rings request \(method) returned NULL"
        case .invalidUtf8: return "Rings returned a non-UTF-8 response"
        case .invalidResponse(let message): return "invalid Rings response: \(message)"
        case .timeout(let message): return "Rings operation timed out: \(message)"
        }
    }
}

public typealias RingsSigner = @Sendable (Data) throws -> Data

private typealias CSigner = @convention(c) (
    UnsafePointer<CChar>?,
    UnsafeMutablePointer<CChar>?
) -> Void
private typealias InitLogging = @convention(c) (Int32) -> Void
private typealias Listen = @convention(c) (UnsafeRawPointer?) -> Void
private typealias Request = @convention(c) (
    UnsafeRawPointer?,
    UnsafePointer<CChar>?,
    UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>?
private typealias StringFree = @convention(c) (UnsafeMutablePointer<CChar>?) -> Void
private typealias ProviderDestroy = @convention(c) (UnsafeMutableRawPointer?) -> Void
private typealias NewProvider = @convention(c) (
    UInt32,
    UnsafePointer<CChar>?,
    UInt64,
    UnsafePointer<CChar>?,
    UnsafePointer<CChar>?,
    CSigner?
) -> UnsafeMutableRawPointer?

/// Dynamically loaded native Rings C ABI.
///
/// Dynamic loading keeps the example independent of Xcode and SwiftPM linker configuration. The
/// path normally points to `target/debug/librings_node.dylib` on macOS or the corresponding `.so`
/// on Linux.
public final class RingsRuntime {
    fileprivate let initLoggingFunction: InitLogging
    fileprivate let listenFunction: Listen
    fileprivate let requestFunction: Request
    fileprivate let stringFreeFunction: StringFree
    fileprivate let providerDestroyFunction: ProviderDestroy
    fileprivate let newProviderFunction: NewProvider
    private let library: UnsafeMutableRawPointer

    public init(libraryPath: String) throws {
        guard let library = dlopen(libraryPath, RTLD_NOW | RTLD_LOCAL) else {
            let message = dlerror().map { String(cString: $0) } ?? "unknown dlopen error"
            throw RingsFfiError.loadLibrary(message)
        }
        do {
            self.initLoggingFunction = try loadSymbol(
                library,
                "rings_node_init_logging",
                as: InitLogging.self
            )
            self.listenFunction = try loadSymbol(
                library,
                "rings_node_listen",
                as: Listen.self
            )
            self.requestFunction = try loadSymbol(
                library,
                "rings_node_request",
                as: Request.self
            )
            self.stringFreeFunction = try loadSymbol(
                library,
                "rings_node_string_free",
                as: StringFree.self
            )
            self.providerDestroyFunction = try loadSymbol(
                library,
                "rings_node_provider_destroy",
                as: ProviderDestroy.self
            )
            self.newProviderFunction = try loadSymbol(
                library,
                "rings_node_new_provider_with_callback",
                as: NewProvider.self
            )
            self.library = library
        } catch {
            dlclose(library)
            throw error
        }
    }

    deinit {
        dlclose(library)
    }

    public func initializeLogging(_ level: RingsLogLevel = .error) {
        initLoggingFunction(level.rawValue)
    }
}

/// Owned provider handle mirroring the Python example's `ProviderHandle`.
///
/// The wrapper retains both the native library and signer callback until `close()` or `deinit`.
/// Up to eight providers may be alive at once because the current C callback has no context
/// pointer; fixed C trampolines keep per-provider Swift closures distinct.
public final class RingsProvider {
    private let runtime: RingsRuntime
    private var handle: UnsafeMutableRawPointer?
    private let signerSlot: Int

    public init(
        runtime: RingsRuntime,
        account: String,
        accountType: String = "eip191",
        networkId: UInt32 = 0,
        iceServer: String = "stun://stun.l.google.com",
        stabilizeInterval: UInt64 = 10,
        signer: @escaping RingsSigner
    ) throws {
        let slot = try signerRegistry.allocate(signer)
        let callback = callbackForSlot(slot)
        let provider = iceServer.withCString { ice in
            account.withCString { account in
                accountType.withCString { accountType in
                    runtime.newProviderFunction(
                        networkId,
                        ice,
                        stabilizeInterval,
                        account,
                        accountType,
                        callback
                    )
                }
            }
        }
        guard let provider else {
            let signerFailure = signerRegistry.takeFailure(slot)
            signerRegistry.release(slot)
            throw signerFailure ?? RingsFfiError.providerCreation
        }
        self.runtime = runtime
        self.handle = provider
        self.signerSlot = slot
    }

    deinit {
        close()
    }

    public func listen() {
        runtime.listenFunction(handle)
    }

    public func request(method: String, json: String) throws -> String {
        let response = method.withCString { method in
            json.withCString { json in
                runtime.requestFunction(handle, method, json)
            }
        }
        guard let response else {
            throw signerRegistry.takeFailure(signerSlot) ?? RingsFfiError.request(method)
        }
        defer { runtime.stringFreeFunction(response) }
        guard let value = String(validatingCString: response) else {
            throw RingsFfiError.invalidUtf8
        }
        return value
    }

    public func requestJson(method: String, object: Any) throws -> Any {
        let data = try JSONSerialization.data(withJSONObject: object)
        guard let json = String(data: data, encoding: .utf8) else {
            throw RingsFfiError.invalidUtf8
        }
        let response = try request(method: method, json: json)
        guard let responseData = response.data(using: .utf8) else {
            throw RingsFfiError.invalidUtf8
        }
        return try JSONSerialization.jsonObject(with: responseData)
    }

    public func nodeDid() throws -> String {
        let result = try requestJson(method: "nodeDid", object: [:])
        guard let object = result as? [String: Any], let did = object["did"] as? String else {
            throw RingsFfiError.invalidResponse("nodeDid has no string did")
        }
        return did
    }

    public func createOffer(remoteDid: String) throws -> String {
        let result = try requestJson(method: "createOffer", object: ["did": remoteDid])
        guard let object = result as? [String: Any], let offer = object["offer"] as? String else {
            throw RingsFfiError.invalidResponse("createOffer has no string offer")
        }
        return offer
    }

    public func answerOffer(_ offer: String) throws -> String {
        let result = try requestJson(method: "answerOffer", object: ["offer": offer])
        guard let object = result as? [String: Any], let answer = object["answer"] as? String else {
            throw RingsFfiError.invalidResponse("answerOffer has no string answer")
        }
        return answer
    }

    public func acceptAnswer(_ answer: String) throws -> Any {
        try requestJson(method: "acceptAnswer", object: ["answer": answer])
    }

    public func listPeers() throws -> [[String: Any]] {
        let result = try requestJson(method: "listPeers", object: [:])
        guard let object = result as? [String: Any],
              let peers = object["peers"] as? [[String: Any]] else {
            throw RingsFfiError.invalidResponse("listPeers has no peer array")
        }
        return peers
    }

    public func peerIsConnected(_ remoteDid: String) throws -> Bool {
        try listPeers().contains { peer in
            peer["did"] as? String == remoteDid && peer["state"] as? String == "Connected"
        }
    }

    public func waitForConnectedPeer(
        _ remoteDid: String,
        timeoutSeconds: TimeInterval = 15,
        pollSeconds: TimeInterval = 0.25
    ) throws {
        try validatePolling(timeoutSeconds: timeoutSeconds, pollSeconds: pollSeconds)
        let deadline = Date().addingTimeInterval(timeoutSeconds)
        while Date() < deadline {
            if try peerIsConnected(remoteDid) {
                return
            }
            Thread.sleep(forTimeInterval: pollSeconds)
        }
        throw RingsFfiError.timeout("peer \(remoteDid) did not reach Connected")
    }

    public func connect(
        to responder: RingsProvider,
        timeoutSeconds: TimeInterval = 15,
        pollSeconds: TimeInterval = 0.25
    ) throws {
        let initiatorDid = try nodeDid()
        let responderDid = try responder.nodeDid()
        let offer = try createOffer(remoteDid: responderDid)
        let answer = try responder.answerOffer(offer)
        _ = try acceptAnswer(answer)
        try waitForConnectedPeer(
            responderDid,
            timeoutSeconds: timeoutSeconds,
            pollSeconds: pollSeconds
        )
        try responder.waitForConnectedPeer(
            initiatorDid,
            timeoutSeconds: timeoutSeconds,
            pollSeconds: pollSeconds
        )
    }

    public func takeE2eEvents() throws -> [[String: Any]] {
        let result = try requestJson(method: "takeE2eEvents", object: [:])
        guard let object = result as? [String: Any],
              let events = object["events"] as? [[String: Any]] else {
            throw RingsFfiError.invalidResponse("takeE2eEvents has no event array")
        }
        return events
    }

    public func sendE2eHandshake(destinationDid: String) throws -> String {
        let result = try requestJson(
            method: "sendE2eHandshake",
            object: ["destination_did": destinationDid]
        )
        guard let object = result as? [String: Any], let txId = object["tx_id"] as? String else {
            throw RingsFfiError.invalidResponse("sendE2eHandshake has no string tx_id")
        }
        return txId
    }

    public func sendE2eMessage(
        destinationDid: String,
        recipientPublicKey: String,
        plaintext: Data,
        maxPlaintextFrameLength: Int = 0
    ) throws -> String {
        let result = try requestJson(
            method: "sendE2eMessage",
            object: [
                "destination_did": destinationDid,
                "recipient_public_key": recipientPublicKey,
                "data": plaintext.base64EncodedString(),
                "max_plaintext_frame_len": maxPlaintextFrameLength,
            ]
        )
        guard let object = result as? [String: Any],
              let streamId = object["stream_id"] as? String else {
            throw RingsFfiError.invalidResponse("sendE2eMessage has no string stream_id")
        }
        return streamId
    }

    public func waitForE2eEvent(
        timeoutSeconds: TimeInterval = 15,
        pollSeconds: TimeInterval = 0.25,
        matching predicate: ([String: Any]) -> Bool
    ) throws -> [String: Any] {
        try validatePolling(timeoutSeconds: timeoutSeconds, pollSeconds: pollSeconds)
        let deadline = Date().addingTimeInterval(timeoutSeconds)
        while Date() < deadline {
            if let event = try takeE2eEvents().first(where: predicate) {
                return event
            }
            Thread.sleep(forTimeInterval: pollSeconds)
        }
        throw RingsFfiError.timeout("matching E2E event was not observed")
    }

    public func waitForE2eStream(
        _ streamId: String,
        timeoutSeconds: TimeInterval = 15,
        pollSeconds: TimeInterval = 0.25
    ) throws -> [[String: Any]] {
        try validatePolling(timeoutSeconds: timeoutSeconds, pollSeconds: pollSeconds)
        var frames: [[String: Any]] = []
        let deadline = Date().addingTimeInterval(timeoutSeconds)
        while Date() < deadline {
            frames.append(contentsOf: try takeE2eEvents().filter { event in
                event["kind"] as? String == "streamFrame"
                    && event["stream_id"] as? String == streamId
            })
            if frames.contains(where: { $0["is_final"] as? Bool == true }) {
                return frames
            }
            Thread.sleep(forTimeInterval: pollSeconds)
        }
        throw RingsFfiError.timeout("E2E stream \(streamId) did not reach its final frame")
    }

    private func validatePolling(timeoutSeconds: TimeInterval, pollSeconds: TimeInterval) throws {
        guard timeoutSeconds > 0, pollSeconds > 0 else {
            throw RingsFfiError.invalidResponse("poll timeout and interval must be positive")
        }
    }

    public func close() {
        guard let handle else { return }
        self.handle = nil
        runtime.providerDestroyFunction(handle)
        signerRegistry.release(signerSlot)
    }
}

private func loadSymbol<T>(
    _ library: UnsafeMutableRawPointer,
    _ name: String,
    as type: T.Type
) throws -> T {
    guard let symbol = dlsym(library, name) else {
        throw RingsFfiError.loadSymbol(name)
    }
    return unsafeBitCast(symbol, to: type)
}

private final class SignerRegistry: @unchecked Sendable {
    private let lock = NSLock()
    private var signers: [Int: RingsSigner] = [:]
    private var failures: [Int: Error] = [:]

    func allocate(_ signer: @escaping RingsSigner) throws -> Int {
        lock.lock()
        defer { lock.unlock() }
        for slot in 0..<8 where signers[slot] == nil {
            signers[slot] = signer
            return slot
        }
        throw RingsFfiError.signerCapacity
    }

    func release(_ slot: Int) {
        lock.lock()
        signers.removeValue(forKey: slot)
        failures.removeValue(forKey: slot)
        lock.unlock()
    }

    func sign(_ slot: Int, message: Data) -> Result<Data, Error> {
        lock.lock()
        let signer = signers[slot]
        lock.unlock()
        guard let signer else { return .failure(RingsFfiError.signerCapacity) }
        do {
            let signature = try signer(message)
            guard signature.count == ringsSignatureLength else {
                return .failure(RingsFfiError.signerLength(signature.count))
            }
            return .success(signature)
        } catch {
            return .failure(error)
        }
    }

    func recordFailure(_ slot: Int, _ error: Error) {
        lock.lock()
        failures[slot] = error
        lock.unlock()
    }

    func takeFailure(_ slot: Int) -> Error? {
        lock.lock()
        defer { lock.unlock() }
        return failures.removeValue(forKey: slot)
    }
}

private let signerRegistry = SignerRegistry()

private func dispatchSigner(
    slot: Int,
    message: UnsafePointer<CChar>?,
    output: UnsafeMutablePointer<CChar>?
) {
    guard let message, let output else { return }
    let input = Data(bytes: message, count: strlen(message))
    switch signerRegistry.sign(slot, message: input) {
    case .success(let signature):
        signature.withUnsafeBytes { bytes in
            if let source = bytes.baseAddress {
                memcpy(output, source, ringsSignatureLength)
            }
        }
    case .failure(let error):
        memset(output, 0, ringsSignatureLength)
        signerRegistry.recordFailure(slot, error)
    }
}

private let signerCallback0: CSigner = { dispatchSigner(slot: 0, message: $0, output: $1) }
private let signerCallback1: CSigner = { dispatchSigner(slot: 1, message: $0, output: $1) }
private let signerCallback2: CSigner = { dispatchSigner(slot: 2, message: $0, output: $1) }
private let signerCallback3: CSigner = { dispatchSigner(slot: 3, message: $0, output: $1) }
private let signerCallback4: CSigner = { dispatchSigner(slot: 4, message: $0, output: $1) }
private let signerCallback5: CSigner = { dispatchSigner(slot: 5, message: $0, output: $1) }
private let signerCallback6: CSigner = { dispatchSigner(slot: 6, message: $0, output: $1) }
private let signerCallback7: CSigner = { dispatchSigner(slot: 7, message: $0, output: $1) }

private func callbackForSlot(_ slot: Int) -> CSigner {
    switch slot {
    case 0: return signerCallback0
    case 1: return signerCallback1
    case 2: return signerCallback2
    case 3: return signerCallback3
    case 4: return signerCallback4
    case 5: return signerCallback5
    case 6: return signerCallback6
    default: return signerCallback7
    }
}

File: examples/ffi/kotlin/src/main/kotlin/network/rings/ffi/Rings.kt
package network.rings.ffi

import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.NativeLibrary
import com.sun.jna.Pointer
import java.io.Closeable
import java.nio.charset.StandardCharsets
import java.util.Base64
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

const val RINGS_SIGNATURE_LENGTH: Int = 65

enum class RingsLogLevel(val abiValue: Int) {
    DEBUG(0),
    INFO(1),
    WARN(2),
    ERROR(3),
    TRACE(4),
}

fun interface RingsSigner {
    fun sign(message: ByteArray): ByteArray
}

internal fun interface SignerCallback : Callback {
    fun invoke(message: Pointer?, output: Pointer?)
}

internal interface NativeRings : Library {
    fun rings_node_init_logging(level: Int)
    fun rings_node_listen(provider: Pointer?)
    fun rings_node_request(provider: Pointer?, method: String, params: String): Pointer?
    fun rings_node_string_free(value: Pointer?)
    fun rings_node_provider_destroy(provider: Pointer?)
    fun rings_node_new_provider_with_callback(
        networkId: Int,
        iceServer: String,
        stabilizeInterval: Long,
        account: String,
        accountType: String,
        signer: SignerCallback,
    ): Pointer?
}

class RingsFfiException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

/** Dynamically loaded desktop Rings C ABI. */
class RingsRuntime private constructor(private val native: NativeRings) {
    companion object {
        private val requiredSymbols = listOf(
            "rings_node_init_logging",
            "rings_node_listen",
            "rings_node_request",
            "rings_node_string_free",
            "rings_node_provider_destroy",
            "rings_node_new_provider_with_callback",
        )

        fun load(libraryPath: String): RingsRuntime {
            val library = NativeLibrary.getInstance(libraryPath)
            requiredSymbols.forEach(library::getFunction)
            return RingsRuntime(Native.load(libraryPath, NativeRings::class.java))
        }
    }

    fun initializeLogging(level: RingsLogLevel = RingsLogLevel.ERROR) {
        native.rings_node_init_logging(level.abiValue)
    }

    fun createProvider(
        account: String,
        signer: RingsSigner,
        accountType: String = "eip191",
        networkId: Int = 0,
        iceServer: String = "stun://stun.l.google.com",
        stabilizeInterval: Long = 10,
    ): RingsProvider = RingsProvider(
        native,
        account,
        signer,
        accountType,
        networkId,
        iceServer,
        stabilizeInterval,
    )
}

/** Owned provider handle aligned with `examples/ffi/rings.py`. */
class RingsProvider internal constructor(
    private val native: NativeRings,
    account: String,
    signer: RingsSigner,
    accountType: String,
    networkId: Int,
    iceServer: String,
    stabilizeInterval: Long,
) : Closeable {
    private val signerFailure = AtomicReference<Throwable?>()
    private val callback = SignerCallback { message, output ->
        try {
            require(message != null && output != null) { "Rings passed a NULL signer pointer" }
            val signature = signer.sign(message.getString(0).toByteArray(StandardCharsets.UTF_8))
            require(signature.size == RINGS_SIGNATURE_LENGTH) {
                "signer returned ${signature.size} bytes; expected $RINGS_SIGNATURE_LENGTH"
            }
            output.write(0, signature, 0, signature.size)
        } catch (error: Throwable) {
            output?.clear(RINGS_SIGNATURE_LENGTH.toLong())
            signerFailure.set(error)
        }
    }
    private var handle: Pointer? = native.rings_node_new_provider_with_callback(
        networkId,
        iceServer,
        stabilizeInterval,
        account,
        accountType,
        callback,
    ) ?: run {
        val signerError = signerFailure.getAndSet(null)
        throw RingsFfiException(
            "rings_node_new_provider_with_callback returned NULL",
            signerError,
        )
    }

    fun listen() {
        native.rings_node_listen(requireHandle())
    }

    fun request(method: String, json: String): String {
        val response = native.rings_node_request(requireHandle(), method, json)
        if (response == null) {
            val signerError = signerFailure.getAndSet(null)
            throw RingsFfiException("Rings request $method returned NULL", signerError)
        }
        return try {
            response.getString(0)
        } finally {
            native.rings_node_string_free(response)
        }
    }

    fun requestJson(method: String, json: String): JsonElement = JSON.parseToJsonElement(
        request(method, json),
    )

    fun nodeDid(): String = requestJson("nodeDid", "{}").jsonObject.string("did")

    fun createOffer(remoteDid: String): String = requestJson(
        "createOffer",
        buildJsonObject { put("did", remoteDid) }.toString(),
    ).jsonObject.string("offer")

    fun answerOffer(offer: String): String = requestJson(
        "answerOffer",
        buildJsonObject { put("offer", offer) }.toString(),
    ).jsonObject.string("answer")

    fun acceptAnswer(answer: String): JsonElement = requestJson(
        "acceptAnswer",
        buildJsonObject { put("answer", answer) }.toString(),
    )

    fun listPeers(): List<JsonObject> = requestJson("listPeers", "{}")
        .jsonObject.array("peers")
        .map { it.jsonObject }

    fun peerIsConnected(remoteDid: String): Boolean = listPeers().any { peer ->
        peer.stringOrNull("did") == remoteDid && peer.stringOrNull("state") == "Connected"
    }

    fun waitForConnectedPeer(
        remoteDid: String,
        timeoutMillis: Long = 15_000,
        pollMillis: Long = 250,
    ) {
        poll(timeoutMillis, pollMillis, "peer $remoteDid did not reach Connected") {
            peerIsConnected(remoteDid)
        }
    }

    fun connectTo(
        responder: RingsProvider,
        timeoutMillis: Long = 15_000,
        pollMillis: Long = 250,
    ) {
        val initiatorDid = nodeDid()
        val responderDid = responder.nodeDid()
        val offer = createOffer(responderDid)
        val answer = responder.answerOffer(offer)
        acceptAnswer(answer)
        waitForConnectedPeer(responderDid, timeoutMillis, pollMillis)
        responder.waitForConnectedPeer(initiatorDid, timeoutMillis, pollMillis)
    }

    fun takeE2eEvents(): List<JsonObject> = requestJson("takeE2eEvents", "{}")
        .jsonObject.array("events")
        .map { it.jsonObject }

    fun sendE2eHandshake(destinationDid: String): String = requestJson(
        "sendE2eHandshake",
        buildJsonObject { put("destination_did", destinationDid) }.toString(),
    ).jsonObject.string("tx_id")

    fun sendE2eMessage(
        destinationDid: String,
        recipientPublicKey: String,
        plaintext: ByteArray,
        maxPlaintextFrameLength: Int = 0,
    ): String = requestJson(
        "sendE2eMessage",
        buildJsonObject {
            put("destination_did", destinationDid)
            put("recipient_public_key", recipientPublicKey)
            put("data", Base64.getEncoder().encodeToString(plaintext))
            put("max_plaintext_frame_len", maxPlaintextFrameLength)
        }.toString(),
    ).jsonObject.string("stream_id")

    fun waitForE2eEvent(
        timeoutMillis: Long = 15_000,
        pollMillis: Long = 250,
        predicate: (JsonObject) -> Boolean,
    ): JsonObject {
        var match: JsonObject? = null
        poll(timeoutMillis, pollMillis, "matching E2E event was not observed") {
            match = takeE2eEvents().firstOrNull(predicate)
            match != null
        }
        return match ?: throw RingsFfiException("matching E2E event disappeared")
    }

    fun waitForE2eStream(
        streamId: String,
        timeoutMillis: Long = 15_000,
        pollMillis: Long = 250,
    ): List<JsonObject> {
        val frames = mutableListOf<JsonObject>()
        poll(timeoutMillis, pollMillis, "E2E stream $streamId did not reach its final frame") {
            frames += takeE2eEvents().filter { event ->
                event.stringOrNull("kind") == "streamFrame"
                    && event.stringOrNull("stream_id") == streamId
            }
            frames.any { event -> event["is_final"]?.jsonPrimitive?.content == "true" }
        }
        return frames
    }

    override fun close() {
        val provider = handle ?: return
        handle = null
        native.rings_node_provider_destroy(provider)
    }

    private fun requireHandle(): Pointer = handle ?: throw RingsFfiException("provider is closed")
}

private val JSON = Json {
    ignoreUnknownKeys = true
}

private fun JsonObject.string(field: String): String = this[field]?.jsonPrimitive?.content
    ?: throw RingsFfiException("response does not contain string field $field")

private fun JsonObject.stringOrNull(field: String): String? = this[field]?.jsonPrimitive?.content

private fun JsonObject.array(field: String): JsonArray = this[field]?.jsonArray
    ?: throw RingsFfiException("response does not contain array field $field")

private fun poll(
    timeoutMillis: Long,
    pollMillis: Long,
    timeoutMessage: String,
    complete: () -> Boolean,
) {
    if (timeoutMillis <= 0 || pollMillis <= 0) {
        throw RingsFfiException("poll timeout and interval must be positive")
    }
    val deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMillis)
    while (System.nanoTime() < deadline) {
        if (complete()) return
        try {
            Thread.sleep(pollMillis)
        } catch (error: InterruptedException) {
            Thread.currentThread().interrupt()
            throw RingsFfiException("polling was interrupted", error)
        }
    }
    throw RingsFfiException(timeoutMessage)
}
